src/shared_utilities/implicit_timestepping.jl

```diff
@@ -0,0 +1,166 @@
+using ClimaCore.MatrixFields
+import ClimaCore.MatrixFields: @name, ⋅
+import UnrolledUtilities
+import LinearAlgebra
+import LinearAlgebra: I
+
+export make_tendency_jacobian,
+    make_update_jacobian, set_dfluxBCdY!, ImplicitEquationJacobian
+
+
+"""
+   make_tendency_jacobian(model::AbstractModel)
+
+Creates and returns a function which updates the auxiliary
+variables `p` in place and then updates the entries of the
+Jacobian matrix `W` for the `model` in place.
+
+The default is that no updates are required, no implicit tendency is
+present, and hence the timestepping is entirely explicit.
+
+Note that the returned function `tendency_jacobian!` should be
+used as `Wfact!` in `ClimaTimeSteppers.jl` and `SciMLBase.jl`.
+"""
+function make_tendency_jacobian(model::AbstractModel)
+    update_aux! = make_update_aux(model)
+    update_boundary_fluxes! = make_update_boundary_fluxes(model)
+    update_jacobian! = make_update_jacobian(model)
+    function tendency_jacobian!(W, Y, p, dtγ, t)
+        update_aux!(p, Y, t)
+        update_boundary_fluxes!(p, Y, t)
+        update_jacobian!(W, Y, p, dtγ, t)
+    end
+    return tendency_jacobian!
+end
+
+"""
+    make_update_jacobian(model::AbstractModel)
+
+Creates and returns a function which updates the entries
+of the Jacobian matrix `W` in place.
+
+If the implicit tendency function is given by
+`T!(dY, Y, p, t) = make_implicit_tendency(model)`, the Jacobian
+should be given by `W_{i,j}! = ∂T!_i/∂Y_j`, where `Y_j` is the
+`j-th` state variable
+and `T!_i` is the implicit tendency of the `i-th` state variable.
+
+The default is that no updates are required, no implicit tendency is
+present, and hence the timestepping is entirely explicit.
+"""
+function make_update_jacobian(model::AbstractModel)
+    function update_jacobian!(W, Y, p, dtγ, t) end
```

Comment:
Should this do nothing? or should it return minus identity matrix? 

or do we never rely on this default (and never will)? if so, maybe we can note that this is only a stub and must be extended or ? not sure.

Reply:
I don't think it needs to return -I. I know that we had previously set the jacobian blocks for explicitly-stepped variables to -I in `update_jacobian!`, but I realized that we just can initialize these as -I in our `ImplicitEquationJacobian` constructor. Then we just do nothing for these blocks in `update_jacobian!`.

And now that I'm writing this out, I notice that `update_jacobian!` is now identical for `RichardsModel` and `EnergyHydrology` with the jacobian approximation we're using right now. I can move it into `src/Soil` and change the type dispatch to use one method for both, or we could just keep it split up since we'll be changing the `EnergyHydrology` version as we iterate on the jacobian anyway. I'm fine with either.

I'll add a note to the generic `update_jacobian!` to explain that it needs to be extended